### PR TITLE
Fixed #29338 -- Allowed using combined queryset in Subquery.

### PR DESCRIPTION
**Pull request by felixxm**
https://code.djangoproject.com/ticket/29338

### CODE REVIEW

File: tests/queries/test_qs_combinators.py [hunk 1]
```diff
@@ -177,6 +177,41 @@ def test_union_with_values_list_and_order(self):
...
+            ExtraInfo(value=5, info='e2'),
+            ExtraInfo(value=7, info='e3'),
+        ])
+        Author.objects.bulk_create([
+            Author(name='a1', num=1, extra=info[2]),
+            Author(name='a2', num=3, extra=info[1]),
+            Author(name='a3', num=2, extra=info[0]),
+        ])
+        qs1 = ExtraInfo.objects.order_by().filter(value=OuterRef('num'))
+        qs2 = ExtraInfo.objects.order_by().filter(value__lt=OuterRef('extra__value'))
```
Comment (felixxm) [inline tests/queries/test_qs_combinators.py]: In this case `extra_value` resolves incorrectly to
```
WHERE "queries_extrainfo"."value" < "queries_extrainfo"."value"
```

**Comment by kaduuuken** [comment]
@felixxm What is the status here? I'm currently running into this issue which causes a lot of performance issues, since have to combine Annotations otherwise...

**Comment by felixxm** [comment]
@kaduuuken We have proper tests and a draft fix. Unfortunately `test_union_in_subquery_related_outerref` fails because [`extra_value` resolves incorrectly](https://github.com/django/django/pull/11692#discussion_r316051867). Please, feel-free to work on this patch.

File: tests/queries/test_qs_combinators.py (hunk 1 above)
Comment (charettes) [reply tests/queries/test_qs_combinators.py]: Does it happen even `union` is not used?

Comment (felixxm) [reply tests/queries/test_qs_combinators.py]: No, without `union` it resolves properly :thinking: .

Comment (ChetanKhanna) [reply tests/queries/test_qs_combinators.py]: Well, am sorry for being too naive. But does this sql seems correct?

`SELECT "queries_author"."name" 
FROM "queries_author" 
INNER JOIN "queries_extrainfo" 
ON ("queries_author"."extra_id" = "queries_extrainfo"."id") 
WHERE (SELECT U0."info" 
FROM "queries_extrainfo" U0 
WHERE U0."value" = "queries_author"."num" 
UNION SELECT U0."info" 
FROM "queries_extrainfo" U0 
WHERE U0."value" < "queries_extrainfo"."value") 
IS NOT NULL 
ORDER BY "queries_author"."name" ASC
`

If so, I got this by inserting an extra `bump_prefix` call for the `combined_query`. Maybe I can open a PR for further investigation?

Comment (felixxm) [reply tests/queries/test_qs_combinators.py]: Thanks for an idea :+1: I also added `combined_queries` to `clone()` to prevent mutations.

**Comment by felixxm** [comment]
@charettes Can you take a look? :pray:

**Comment by felixxm** [comment]
Failures on MySQL and MariaDB are related with combining parenthesized queries (see [1](https://mariadb.com/kb/en/union/#parentheses) and [2](https://dev.mysql.com/doc/refman/8.0/en/union.html)). I will try to fix them later.

File: django/db/models/sql/query.py [hunk 2]
```diff
@@ -328,6 +328,8 @@ def clone(self):
             obj.subq_aliases = self.subq_aliases.copy()
         obj.used_aliases = self.used_aliases.copy()
         obj._filtered_relations = self._filtered_relations.copy()
+        if self.combined_queries:
+            obj.combined_queries = copy.deepcopy(obj.combined_queries)
```
Comment (charettes) [inline django/db/models/sql/query.py]: I'd try calling `clone` recursively instead.

```suggestion
            obj.combined_queries = tuple(query.clone() for query in obj.combined_queries)
```

File: django/db/models/sql/query.py [hunk 3]
```diff
@@ -1033,6 +1035,10 @@ def resolve_expression(self, query, *args, **kwargs):
                 not self.select_for_update):
             clone.clear_ordering(True)
         clone.where.resolve_expression(query, *args, **kwargs)
+        # Resolve combined queries.
+        for combined_query in clone.combined_queries:
+            combined_query.bump_prefix(query)
+            combined_query.where.resolve_expression(query, *args, **kwargs)
```
Comment (charettes) [inline django/db/models/sql/query.py]: I'm curious as to whether calling `combined_query.resolve_expression(query, *args, **kwargs)` instead to trigger a recursive resolving would work here?

```python
clone.combined_queries = tuple(
    query.resolve_expression(query, *args, **kwargs) for query in clone.combined_queries
)
```

Comment (ChetanKhanna) [reply django/db/models/sql/query.py]: I can confirm that this works :+1:

File: django/db/models/sql/query.py (hunk 2 above)
Comment (ChetanKhanna) [reply django/db/models/sql/query.py]: If I'm not mistaken we won't need this at all after the changes suggested below.

Comment (charettes) [reply django/db/models/sql/query.py]: @ChetanKhanna yeah I agree that it's likely unnecessary given `tuples` are immutable and we don't alter `combined_queries` except during expression resolving.

**Comment by felixxm** [comment]
@charettes Thanks for ideas :+1: , I will implement them tomorrow.

[Parenthesized combined queries](https://github.com/django/django/blob/master/django/db/models/sql/compiler.py#L471) raise syntax error on [MariaDB < 10.4](https://mariadb.com/kb/en/union/#parentheses) and [MySQL < 8.0](https://dev.mysql.com/worklog/task/?id=8083), I need to find out when/where we need them.

**Comment by charettes** [comment]
@felixxm 

> Parenthesized combined queries raise syntax error on MariaDB < 10.4 and MySQL < 8.0, I need to find out when/where we need them.

Maybe you'll be able to branch of `self.query.subquery` but that's just a shot in the dark, I haven't experienced with MySQL combined subqueries in the past.

**Comment by ChetanKhanna** [comment]
> [Parenthesized combined queries](https://github.com/django/django/blob/master/django/db/models/sql/compiler.py#L471) raise syntax error on [MariaDB < 10.4](https://mariadb.com/kb/en/union/#parentheses) and [MySQL < 8.0](https://dev.mysql.com/worklog/task/?id=8083), I need to find out when/where we need them.

@felixxm I took the liberty to look into this and was wondering if it would be sufficient to turn `supports_slicing_ordering_in_compound` in `mysql.features` into a property so as to check for `connection.mysql_version`? I wasn't able test this against the entire test suite with mysql though.

**Comment by felixxm** [comment]
@ChetanKhanna It's not strictly related with supporting slicing ordering in compound, this flag is set properly for supported versions of MySQL and MariaDB. I have to consider all cases. I should find some time in the next few weeks.

**Comment by sjamaan** [comment]
What needs to be done for this to land in the next release?  I have several ORM-generated queries that can be sped up enormously by adding an outerref to a subselect which is a union and would love to be able to use it.

**Comment by felixxm** [comment]
@sjamaan We have few months to the next release. I should be able to finish this at that time.

**Comment by knyghty** [comment]
Has this fallen off the radar? I find myself with a use-case for this.

**Comment by insspb** [comment]
yep, need this too.

**Comment by auvipy** [comment]
> @sjamaan We have few months to the next release. I should be able to finish this at that time.

would love to see this in 4.1 O:) of course if you have time

**Comment by felixxm** [comment]
It seems that I've finally found a way to make it work on all databases. Without the last commit (_"Edits"_) it crashes on:
- [MariaDB < 10.4](https://mariadb.com/kb/en/union/#parentheses),
- [MySQL < 8](https://dev.mysql.com/worklog/task/?id=8083).

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Review by charettes — APPROVED** [review]
The adjustments to have `get_combinator_sql` branch off `self.subquery` make sense to me.

**Comment by felixxm** [comment]
@charettes Thanks :+1:

**Pull request merged**